text/0000-stable-rustdoc-urls.md

```diff
@@ -0,0 +1,260 @@
+
+## Re-exports will generate a page pointing to the canonical version
+
+The redirect page will go in the same place as the re-export would be if it
+were inlined with `#[doc(inline)]` after this RFC.
+
+There will _not_ be a page generated at `kind.name.html` at the level of the re-export, since it's not possible for there to be any existing links there that were not broken.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Rust is case-sensitive, but some filesystems (especially on Windows) are not,
+  so there are naming collisions in the files Rustdoc generates ([#25879]).
+  If Rustdoc combines several 'kinds' into one namespace, there will be more conflicts than currently:
+
+```rust
+struct Command; // page generated at `t.Command.html`
+enum command {} // page generated at `t.command.html`
+```
+
+**@nemo157** has kindly conducted a survey of the docs.rs documentation and found
+that there are about 700,000 items that [currently overlap]. After this change,
+that would go up to about [850,000 items that overlap][overlap-after-change].
+docs.rs has 308,064,859 total items in the inventory, so previously 0.23% files conflicted
+and after this RFC 0.28% files will conflict.
+
+In the opinion of the author, since this is an existing problem,
+it does not need to be solved in order to go forward with the RFC.
+
+[currently overlap]: https://ipfs.io/ipfs/QmfZatebkVFdEUYQtPaAitsBbmLKAtQgaWLciSnjtLAWfv/case-conflicts.txt.zst
+[overlap-after-change]: https://ipfs.io/ipfs/QmfZatebkVFdEUYQtPaAitsBbmLKAtQgaWLciSnjtLAWfv/cased-namespace-conflicts.txt.zst
+
+[#25879]: https://github.com/rust-lang/rust/issues/25879
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+## How were the URLs chosen?
+
+There were three main criteria for choosing the URLs (in vauge order of priority):
+
+1. They should be based on the namespace, not the 'kind' of the item. Otherwise there's not much point to the RFC, because the URLs won't be stable.
+2. They should make sense when viewed; for example `a`, `b`, `c` would be bad choices for the names.
+3. They should be fairly short, so they're easy to type; for example `type_namespace.` would not be a great choice.
+
+`t.` and `m.` were partly chosen based on precedent in [#35236] (but see 'Naming alternatives' below for the main reason).
+
+### Naming alternatives
+
+Note that these names are easy to 'bikeshed' and don't substantially change the RFC.
+
+- Rustdoc could remove the `v.` prefix for items in the value namespace.
+  This would make the URLs for functions slightly less confusing, but introduce a conflict for functions named `index()`, since rustdoc has to generate `index.html` for modules.
+- Rustdoc could lengthen the prefixes to `value.`, `type.` and `macro.`. This makes the URLs easier to read, at the cost of making them more confusing for traits (consider `type.Trait.html`).
```

Comment:
> at the cost of making them more confusing for traits

I'd say this same argument could be provided against the short version `t.Trait.html` too. What does even mean `t.` here? Unless I read the current RFC, it's going to be confusing too, isn't it?

It feels to me we could list the pros and cons for the short and the long version:
- pro-short: short URL, less typing
- cons-short: confusing
- pro-long: explicit
- cons-long: more typing, confusing

And about _typing_, I'm not even sure it's a valid argument **for** _short_ or **against** _long_ because most of it is going to be automated in RustDoc.

What was the rationale behind choosing the _short_ version over the _long_ version besides the supposed confusing aspect?

Reply:
`t.` is intentionally ambiguous - it means 'type namespace', but you could imagine it also meaning 'trait'. Like you said, it will be automated in rustdoc.

What benefits do you see for making it the long version instead?

Reply:
Since the computer (aka RustDoc in this case) is doing it for me, I usually prefer long names because:
- I don't have to type them so length doesn't matter
- it is self-documented, meaning anyone who sees it can understand the prefix (whereas the short version is not as obvious)

Reply:
Hmm, do you think `type.Trait.html` is less confusing than `t.Trait.html`? @elidupree pointed that out as a drawback of the other approach in https://internals.rust-lang.org/t/pre-rfc-stable-rustdoc-urls/13099/15.

Reply:
First of all, thank you to take the time trying to understand my position.

I do not think `type.Trait.html` is **too** confusing. But this is only my opinion.

My point is: if `type.Trait.html` is confusing, then I don't understand how `t.Trait.html` is less confusing since it does mean the same, just in an abbreviated way. And this is just a nice coincidence that the first letter of _type_ happens to be the same as _trait_... but playing that card is only trumping people.

(please excuse if the tone is perceived too harsh, English not being my first language, I do not intend to be harsh :smile:)

For me, I prefer long names for the same reason I put `--output` in a bash script instead of `-o`: both achieved the same functionality but for one of them, I don't have to go look the documentation to understand what it is.

Reply:
Ok, I changed it back to `type.` @elidupree if you still have concerns please speak up :)

Reply:
I'll repeat what I said in the IRLO thread:

The scenario I'm imagining is that someone will be looking for a trait, click to the trait page, glance at the URL bar, think "oh wait, I was looking for a trait but I accidentally clicked to a *type*", and go back to look elsewhere for the trait.

This won't happen *that* often, but when it does, it's a sizable mental stumbling block for someone who is trying to just use the documentation as intended. The `t.Trait.html` naming scheme avoids this problem.

(Having separate `type.Name.html` and `trait.Name.html` entries would also avoid the problem, and I still think that would be a reasonable approach, since it's hard to imagine a scenario where you'd try to replace a type with a trait in a semver-compatible way. But if we're using the namespace-only approach, `t.` is preferable for this reason.)

Reply:
I agree, most rust users aren't well aware of the namespaces